util/url-encoding/src/url.rs

```diff
@@ -0,0 +1,128 @@
+//! Representation of a mob-scheme url, based on rust-url lib
+//!
+//! This does not follow the pattern of mc-util-uri because this url is not
+//! actually used as the basis of a connection to a service, it's rather an
+//! encoding that represents a public address, or a payment request following
+//! the url specification. It is intended to be more or less human readable
+//! and could be used e.g. in an email where a user might actually look at the
+//! url, and could be registered with the OS so that a mobilecoin client launches
+//! to handle the url.
+//!
+//! Because we don't implement ConnectionUri, don't really have a notion of port,
+//! and aren't required to actually have a hostname (fog-less mob url's don't),
+//! we don't really want to base this on the mc-util-uri type.
+
+use core::{
+    fmt::{Display, Formatter, Result as FmtResult},
+    str::FromStr,
+};
+
+pub use mc_util_uri::UriParseError;
+use url::Url;
+
+pub const MOB_SCHEME_SECURE: &str = "mob";
+pub const MOB_SCHEME_INSECURE: &str = "insecure-mob";
+
+pub const REPORT_KEY_KEY: &str = "k";
+pub const SIG_KEY: &str = "s";
+const AMOUNT_KEY: &str = "a";
+const MEMO_KEY: &str = "m";
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct MobUrl {
+    /// The original Url object used to construct this object.
+    url: Url,
+    /// Whether to use TLS when connecting.
+    use_tls: bool,
+}
+
+impl MobUrl {
+    pub fn url(&self) -> &Url {
+        &self.url
+    }
+
+    pub fn use_tls(&self) -> bool {
+        self.use_tls
+    }
+
+    // query parameters
+    fn get_param(&self, name: &str) -> Option<String> {
+        self.url().query_pairs().find_map(|(k, v)| {
+            if k == name && !v.is_empty() {
+                Some(v.to_string())
+            } else {
+                None
+            }
+        })
+    }
+
+    // TODO: We should probably have some functionality to eliminate duplicate parameters
```

Comment:
We shouldn't fight the fact that URL params are effectively `HashMap<String, Vec<String>>`. We can implement logic (as you've done) for well-known parameters like `k`, but we shouldn't get ahead of ourselves for the general case.

Reply:
what i mean is, I'm not sure that the rust / servo url lib itself will detect duplicate keys when we add new key value pairs, I think the serializer might just be really dumb. it would be annoying if you start with a fog url that already has `a` or `k` and end up with `a` or `k` appearing twice

Reply:
i still haven't added a test to see what happens here / what should happen here,  i want to do that

Reply:
done